oak/server/storage/storage_processor.cc

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "oak/server/storage/storage_processor.h"
+
+#include "absl/memory/memory.h"
+#include "absl/strings/escaping.h"
+#include "asylo/util/cleansing_types.h"
+#include "asylo/util/logging.h"
+#include "grpcpp/create_channel.h"
+
+namespace oak {
+
+constexpr size_t kMaxMessageSize = 1 << 16;
+
+static std::string GetStorageId(const std::string& storage_name) {
+  // TODO: Generate name-based UUID.
+  return storage_name;
+}
+
+static asylo::CleansingVector<uint8_t> GetStorageEncryptionKey(const std::string& storage_name) {
+  // TODO: Request encryption key from escrow service.
+  std::string encryption_key =
+      absl::HexStringToBytes("c0dedeadc0dedeadc0dedeadc0dedeadc0dedeadc0dedeadc0dedeadc0dedead");
+  return asylo::CleansingVector<uint8_t>(encryption_key.begin(), encryption_key.end());
+}
+
+StorageProcessor::StorageProcessor(const std::string& storage_address)
+    : fixed_nonce_generator_(new oak::FixedNonceGenerator()),
+      datum_name_cryptor_(kMaxMessageSize, fixed_nonce_generator_),
+      datum_value_cryptor_(kMaxMessageSize, new asylo::AesGcmSivNonceGenerator()),
+      storage_service_(oak::Storage::NewStub(
+          grpc::CreateChannel(storage_address, grpc::InsecureChannelCredentials()))) {}
+
+const asylo::StatusOr<std::string> StorageProcessor::EncryptDatum(const std::string& datum,
+                                                                  DatumType datum_type) {
+  // TODO: Replace "foo" with identifier for the encryption key.
+  asylo::CleansingVector<uint8_t> key = GetStorageEncryptionKey("foo");
+  asylo::CleansingString additional_authenticated_data;
+  asylo::CleansingString nonce;
+  asylo::CleansingString datum_encrypted;
+  asylo::AesGcmSivCryptor* cryptor = nullptr;
+
+  switch (datum_type) {
+    case DatumType::NAME: {
+      fixed_nonce_generator_->set_datum_name(datum);
+      cryptor = &datum_name_cryptor_;
+      break;
+    };
+    case DatumType::VALUE: {
+      cryptor = &datum_value_cryptor_;
+      break;
+    };
+  };
```

Comment:
Does C++ check for exhaustiveness when switching on enums? if not, perhaps add a default case throwing an error here.

Reply:
Yes, default case should not be used with enums so that the compiler can check.
https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements